flax/experimental/nnx/nnx/transforms.py

```diff
@@ -133,267 +141,36 @@ class JitStaticOutputs:
-
-  return jitted_fn
+def _jitted_fn(
+  *args: tp.Any,
+  _nnx_jit_static: JitStaticInputs,
+  _nnx_jit_state: State,
+  **kwargs: tp.Any,
+) -> tuple[tp.Any, State, GraphDef[tuple[tp.Any, ...]]]:
+  ctx = _nnx_jit_static.ctx
+  graphdef = _nnx_jit_static.graphdef
+  constrain_state = _nnx_jit_static.constrain_state
+  f = _nnx_jit_static.f
+  state: State = _nnx_jit_state
 
+  if constrain_state is not None:
+    state = constrain_state(state)
 
-def jit_apply(
-  options: JITOptions,
-  jitted_fn: JittedFn,
-  args: tuple[tp.Any, ...],
-  kwargs: dict[str, tp.Any],
-) -> tp.Any:
-  ctx = graph.UpdateContext()
-  (args, kwargs), input_graph_nodes = graph.extract_graph_nodes((args, kwargs))
-  graphdef, state = ctx.split(input_graph_nodes)
+  input_graph_nodes = ctx.merge(graphdef, state)
 
-  out, output_state, output_graphdef = jitted_fn(
-    *args,
-    _nnx_jit_static=JitStaticInputs(graphdef, ctx),
-    _nnx_jit_state=state,
-    **kwargs,
-  )
-  input_graph_nodes, output_graph_nodes = ctx.update(
-    output_graphdef, output_state
-  )
-  out = graph.insert_graph_nodes(out, output_graph_nodes)
-  return out
+  (args, kwargs) = graph.insert_graph_nodes((args, kwargs), input_graph_nodes)
 
+  out = f(*args, **kwargs)
 
-class Jit(tp.Generic[M], LiftedModule[M]):
+  out, output_graph_nodes = graph.extract_graph_nodes(out)
 
-  @staticmethod
-  def constructor(
-      module_constructor: tp.Callable[..., MA],
-      *,
-      in_shardings: tp.Any = UNSPECIFIED,
-      out_shardings: tp.Any = UNSPECIFIED,
-      static_argnums: int | tp.Sequence[int] | None = None,
-      static_argnames: str | tp.Iterable[str] | None = None,
-      donate_argnums: int | tp.Sequence[int] | None = None,
-      donate_argnames: str | tp.Iterable[str] | None = None,
-      keep_unused: bool = False,
-      device: tp.Optional[jax.Device] = None,
-      backend: tp.Optional[str] = None,
-      inline: bool = False,
-      abstracted_axes: tp.Optional[tp.Any] = None,
-      # nnx specific
-      donate_state: bool = False,
-      constrain_state: bool | tp.Callable[[State], State] = False,
-  ) -> tp.Callable[..., 'Jit[MA]']:
-    def _create_jit(*args, **kwargs):
-      return Jit(
-          module_constructor=module_constructor,
-          in_shardings=in_shardings,
-          out_shardings=out_shardings,
-          static_argnums=static_argnums,
-          static_argnames=static_argnames,
-          donate_argnums=donate_argnums,
-          donate_argnames=donate_argnames,
-          keep_unused=keep_unused,
-          device=device,
-          backend=backend,
-          inline=inline,
-          abstracted_axes=abstracted_axes,
-          # nnx specific
-          donate_state=donate_state,
-          constrain_state=constrain_state,
-          # submodule args
-          module_init_args=args,
-          module_init_kwargs=kwargs,
-      )
+  graphdef, state = ctx.split((input_graph_nodes, output_graph_nodes))
+  ctx.clear_idxmap()  # avoid memory leak
```

Comment:
It's not clear from the context why *not* doing this would leak memory. Consider adding a few more details for posterity.